// typescript_worker_threads/utils/lib/Pool.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const Worker_1 = require("./Worker");
const worker_threads_1 = require("worker_threads");
const AggregateError = require("aggregate-error");
const ModuleResolver_1 = require("./ModuleResolver");
class Pool extends events_1.EventEmitter {
    constructor(id, poolSize = 4, presetCallerDirname = null, depth = 5) {
        super();
        this.queue = [];
        this.pool = [];
        this.runningWorkers = 0;
        this.isReady = false;
        this.closing = false;
        this.id = id;
        this.moduleResolver = new ModuleResolver_1.ModuleResolver(this.id, presetCallerDirname, depth);
        this.size = poolSize;
        this.on('worker_online', () => {
            if (this.runningWorkers <= this.size) {
                this.runningWorkers += 1;
                if (this.runningWorkers === this.size) {
                    if (!this.isReady) {
                        this.isReady = true;
                        this.emit('ready');
                    }
                }
            }
        });
        this.on('release', worker => {
            this.onRelease(worker);
        });
        process.nextTick(() => __awaiter(this, void 0, void 0, function* () {
            for (let i = 0; i < this.size; i++) {
                yield this.sleep(1000);
                yield this.asyncAppendPoolWorker();
            }
        }));
    }
    sleep(timeout) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise(resolve => {
                setTimeout(resolve, timeout);
            });
        });
    }
    static handleErrors(maybeErrors) {
        const errors = maybeErrors.filter(error => {
            return error != null;
        });
        if (errors.length === 0) {
            return;
        }
        throw new AggregateError(errors);
    }
    onRelease(worker) {
        if (this.queue.length > 0) {
            const cb = this.queue.shift();
            cb(worker);
        }
        else {
            this.pool.push(worker);
        }
    }
    runWorker(worker, workerData, resolve, reject) {
        const messageChannel = new worker_threads_1.MessageChannel();
        workerData.port = messageChannel.port1;
        messageChannel.port2.once('message', (result) => {
            worker.removeAllListeners('error');
            this.onMessage(resolve, worker, result);
        });
        worker.once('error', (err) => {
            reject(err);
            this.appendPoolWorker();
        });
        worker.postMessage(workerData, [messageChannel.port1]);
    }
    onMessage(resolve, worker, result) {
        this.emit('release', worker);
        resolve(result);
    }
    asyncAppendPoolWorker() {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                try {
                    this.appendPoolWorker(resolve);
                }
                catch (e) {
                    reject(e);
                }
            });
        });
    }
    appendPoolWorker(callback) {
        const jsFilename = this.moduleResolver.resolveModule();
        const isSourceCode = this.moduleResolver.isSourceCode;
        const worker = new Worker_1.Worker(this.id, {
            overwrite: {
                jsFilename: jsFilename,
                isSourceCode: isSourceCode
            }
        });
        worker.once('online', () => {
            if (callback)
                callback();
            this.emit('worker_online');
        });
        worker.once('exit', (exitCode) => {
            if (exitCode !== 0 && !this.closing) {
                this.appendPoolWorker();
            }
        });
        this.pool.push(worker);
    }
    getQueueLength() {
        return this.queue.length;
    }
    getPoolLength() {
        return this.pool.length;
    }
    run(workerData) {
        return new Promise((resolve, reject) => {
            if (this.pool.length > 0) {
                const worker = this.pool.shift();
                this.runWorker(worker, workerData, resolve, reject);
            }
            else {
                this.queue.push(worker => {
                    this.runWorker(worker, workerData, resolve, reject);
                });
            }
        });
    }
    close() {
        this.closing = true;
        const promises = [];
        const size = this.getPoolLength();
        for (let index = 0; index < size; index++) {
            const worker = this.pool.shift();
            promises.push(new Promise((resolve) => {
                worker.terminate().then(resolve);
            }));
        }
        Promise.all(promises)
            .then(maybeErrors => {
            return Pool.handleErrors(maybeErrors);
        })
            .then(() => this.emit('close'))
            .catch((error) => {
            this.emit('error', error);
            this.emit('close');
        });
    }
}
exports.Pool = Pool;
